### PR TITLE
Workaround for ogre crash on shutdown

### PR DESCRIPTION
**Pull request by iche033**
# 🦟 Bug fix

Workaround for https://github.com/gazebosim/gz-rendering/issues/1007

## Summary

The `INTEGRATION_load_unload` test crashes with ogre 1.x. The tests checks to see if the render engine can be loaded and unloaded in a thread (which is how `gz sim` runs the render engine). The crash happens on ubuntu 24.04 with the system debs, see https://github.com/gazebosim/gz-rendering/issues/1007 for more info.

The workaround implemented in this PR is to prevent the `RenderSystem_GL` plugin library from being unloaded on shutdown to avoid the crash. The plugin will be still be stopped and deleted but we do not call `dlclose` on it. I verified that the destructors of various objects / singletons in the plugin are called.

This is not ideal and we should find a proper solution in the future. 

## To test

The `INTEGRATION_load_unload` test should now pass with ogre 1.x on Ubuntu 24.04 with the system ogre debs:

```
GZ_ENGINE_TO_TEST=ogre  ./build/gz-rendering9/bin/INTEGRATION_load_unload
```

gz-sim should no longer crash on exit:

```
gz sim -v 4 -r --iterations 5 -s sensors_demo.sdf --render-engine ogre
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by scpeters** [comment]
potential fix for the remaining cmake warning in https://github.com/gazebo-tooling/release-tools/pull/1161

**Comment by scpeters** [comment]
> potential fix for the remaining cmake warning in [gazebo-tooling/release-tools#1161](https://github.com/gazebo-tooling/release-tools/pull/1161)

testing shows that https://github.com/gazebo-tooling/release-tools/pull/1161 should fix the warning

File: ogre/src/OgreRenderEngine.cc [hunk 1]
```diff
@@ -66,6 +70,84 @@ class gz::rendering::OgreRenderEnginePrivate
 using namespace gz;
 using namespace rendering;
 
+// Plugin loading and unloading code adapted from Ogre.
+// Unloading the RenderSystem_GL plugin was found to cause a crash when the
+// rendering thread exits. This only happens on Ubuntu 24.04 when running
+// ogre using system debs, see
+// https://github.com/gazebosim/gz-rendering/issues/1007
+// To workaournd ths issue, we adapth the code for loadng and unloading
```
Comment (scpeters) [inline ogre/src/OgreRenderEngine.cc:78]:
```suggestion
// To workaournd ths issue, we adapt the code for loadng and unloading
```

**Review by scpeters — COMMENTED** [review]
minor nit; I'm still trying to understand the code

**Comment by azeey** [comment]
We now have API to load plugins with `RTLD_NODELETE`. This would ensure that the plugin is not actually deleted even if `dlclose` is called. Would changing https://github.com/gazebosim/gz-rendering/blob/6a2217fce93cbc9c5db6495c711e6cd9b1c7991d/src/RenderEngineManager.cc#L498 to  `LoadLib(pathToLib, true)` fix this issue without any of the other changes?

File: ogre/src/OgreRenderEngine.cc [hunk 2]
```diff
@@ -118,9 +200,13 @@ void OgreRenderEngine::Destroy()
 
   if (ogreRoot)
   {
+    // TODO(anyone): do we need to catch segfault on delete?
     try
     {
-      // TODO(anyone): do we need to catch segfault on delete?
+      // Install a dummy plugin that allows use to manually shutdown ogre
```
Comment (scpeters) [inline ogre/src/OgreRenderEngine.cc:206]:
```suggestion
      // Install a dummy plugin that allows us to manually shutdown ogre
```

File: ogre/src/OgreRenderEngine.cc [hunk 3]
```diff
@@ -66,6 +70,84 @@ class gz::rendering::OgreRenderEnginePrivate
...
+// rendering thread exits. This only happens on Ubuntu 24.04 when running
+// ogre using system debs, see
+// https://github.com/gazebosim/gz-rendering/issues/1007
+// To workaournd ths issue, we adapth the code for loadng and unloading
+// ogre plugin here. We load the RenderSystem_GL library manually and store it
+// in gz-rendering and on shutdown we stop the plugin and delete the lib without
+// unloading it.
+// \todo(iche033) Find a proper way to unload the library without causing a
+// crash.
+typedef void (*DLL_START_PLUGIN)(void);
```
Comment (scpeters) [inline ogre/src/OgreRenderEngine.cc:84]: `DLL_START_PLUGIN` has a `const NameValuePairList *` argument in [OgreRoot.cpp:132](https://github.com/OGRECave/ogre-next/blob/v2.3.3/OgreMain/src/OgreRoot.cpp#L132), but is `void` here. Does that matter?

File: ogre/src/OgreRenderEngine.cc [hunk 4]
```diff
@@ -66,6 +70,84 @@ class gz::rendering::OgreRenderEnginePrivate
...
+  // on thread exit so commented out for now.
+  // _pluginLib->unload();
+  delete _pluginLib;
+}
+
+/// A dummy plugin class that is installed to the ogre root before shutdown.
+/// We use this as a hook to get a callback from ogre root when all plugins
+/// are being unloaded. This is so that we can unload the RenderSystem_GL
+/// at the right time.
+class DummyPlugin : public Ogre::Plugin
```
Comment (scpeters) [inline ogre/src/OgreRenderEngine.cc:131]: nit: rename to `DummyUnloadPlugin`

**Comment by scpeters** [comment]
> We now have API to load plugins with `RTLD_NODELETE`. This would ensure that the plugin is not actually deleted even if `dlclose` is called. Would changing
> 
> https://github.com/gazebosim/gz-rendering/blob/6a2217fce93cbc9c5db6495c711e6cd9b1c7991d/src/RenderEngineManager.cc#L498
> 
> to `LoadLib(pathToLib, true)` fix this issue without any of the other changes?

testing in a draft PR: https://github.com/gazebosim/gz-rendering/pull/1038

**Comment by scpeters** [comment]
> > We now have API to load plugins with `RTLD_NODELETE`. This would ensure that the plugin is not actually deleted even if `dlclose` is called. Would changing
> > https://github.com/gazebosim/gz-rendering/blob/6a2217fce93cbc9c5db6495c711e6cd9b1c7991d/src/RenderEngineManager.cc#L498
> > 
> > to `LoadLib(pathToLib, true)` fix this issue without any of the other changes?
> 
> testing in a draft PR: #1038

the noble workflow passes; I've marked https://github.com/gazebosim/gz-rendering/pull/1038 as ready for review

File: ogre/src/OgreRenderEngine.cc (hunk 3 above)
Comment (scpeters) [reply ogre/src/OgreRenderEngine.cc:83]: Oops thanks!

**Review by scpeters — APPROVED** [review]
this has fixed CI and looks fine to me; I defer to @azeey about whether to follow the suggestion from @iche033 to use this instead of #1038

**Review by azeey — APPROVED** [review]
I was going to check if windows CI was okay, but it looks like the Ogre1 tests have been failing for a while. I thought it was just Ogre2 that was failing :(. Anyway, I'm good with this approach.

**Comment by iche033** [comment]
yeah we should try setting up ogre 2.3 on windows, https://github.com/gazebosim/gz-rendering/issues/745

**Pull request merged**